lib/Sema/CSDiag.cpp

```diff
@@ -6411,9 +6419,69 @@ static void noteArchetypeSource(const TypeLoc &loc, ArchetypeType *archetype,
   if (!FoundDecl)
     FoundDecl = loc.getType()->getAnyGeneric();
   
-  if (FoundDecl)
+  if (FoundDecl) {
     tc.diagnose(FoundDecl, diag::archetype_declared_in_type, archetype,
                 FoundDecl->getDeclaredType());
+  }
+  if (FoundGenericTypeBase && !isa<GenericIdentTypeRepr>(FoundGenericTypeBase)){
+    assert(FoundDecl);
+
+    SmallString<64> genericParamBuf;
```

Comment:
Is this in any way similar to diagnoseUnboundGenericType() in TypeCheckType.cpp?